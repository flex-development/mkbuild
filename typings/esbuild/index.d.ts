import type WithUndefined from '#types/with-undefined'
import type { Tsconfig } from '@flex-development/tsconfig-types'

declare module 'esbuild' {
  interface TransformOptions {
    tsconfigRaw?: Tsconfig | TsconfigRaw | string
  }

  /**
   * Transform `code`.
   *
   * This function can be used to minify JavaScript, convert TypeScript/JSX to
   * JavaScript, or convert newer JavaScript to older JavaScript.
   *
   * @see {@linkcode TransformOptions}
   * @see {@linkcode TransformResult}
   *
   * @template {WithUndefined<TransformOptions>} T
   *  Transform options
   *
   * @async
   *
   * @param {Uint8Array | string} code
   *  The code to transform
   * @param {SameShape<TransformOptions, T> | undefined} [options]
   *  Transform options
   * @return {Promise<TransformResult<T>>}
   *  Transform result
   */
  export declare function transform<T extends WithUndefined<TransformOptions>>(
    code: Uint8Array | string,
    options?: SameShape<TransformOptions, T>
  ): Promise<TransformResult<T>>
}
