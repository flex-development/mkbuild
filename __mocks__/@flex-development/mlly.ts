/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
