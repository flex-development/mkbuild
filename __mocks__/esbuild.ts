/**
 * @file Mocks - esbuild
 * @module mocks/esbuild
 * @see https://esbuild.github.io
 */

/**
 * [`esbuild`][1] module type.
 *
 * [1]: https://esbuild.github.io
 */
type Actual = typeof import('esbuild')

/**
 * `esbuild` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('esbuild')

export const context = vi.fn(actual.context)
export const transform = vi.fn(actual.transform)
