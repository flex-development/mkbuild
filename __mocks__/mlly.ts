/**
 * @file Mocks - mlly
 * @module mocks/mlly
 * @see https://github.com/unjs/mlly
 */

/**
 * [`mlly`][1] module type.
 *
 * [1]: https://github.com/unjs/mlly
 */
type Actual = typeof import('mlly')

/**
 * `mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('mlly')

export const findDynamicImports = vi.fn(actual.findDynamicImports)
export const findExports = vi.fn(actual.findExports)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const resolvePath = vi.fn(actual.resolvePath)
