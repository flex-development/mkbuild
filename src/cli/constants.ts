/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
