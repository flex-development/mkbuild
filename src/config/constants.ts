/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
