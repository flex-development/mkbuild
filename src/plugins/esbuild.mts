/**
 * @file Plugins - esbuild
 * @module mkbuild/plugins/esbuild
 */

import formatEsbuildMessage from '#internal/format-esbuild-message'
import type WithUndefined from '#types/with-undefined'
import type {
  EsbuildOptions,
  Format
} from '@flex-development/mkbuild'
import pathe from '@flex-development/pathe'
import { ifelse, isNIL } from '@flex-development/tutils'
import { createFilter } from '@rollup/pluginutils'
import { ok } from 'devlop'
import type * as Esbuild from 'esbuild'
import type { Loader, LogLevel, TransformOptions, TsconfigRaw } from 'esbuild'
import type {
  MinimalPluginContext,
  ModuleInfo,
  NormalizedInputOptions,
  NormalizedOutputOptions,
  OutputBundle,
  Plugin,
  PluginContext,
  ProgramNode,
  RenderedChunk,
  RollupLog,
  LogLevel as RollupLogLevel,
  SourceDescription,
  TransformPluginContext,
  TransformResult
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:esbuild'

/**
 * esbuild module.
 *
 * @internal
 *
 * @var {typeof Esbuild | undefined} esbuild
 */
let esbuild: typeof Esbuild | undefined

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Plugin options.
 *
 * @internal
 *
 * @extends {EsbuildOptions}
 */
interface Options extends EsbuildOptions {
  /**
   * Module format.
   *
   * @see {@linkcode Format}
   */
  format: Format

  /**
   * Set the value of the `sourceRoot` field in a source map, which specifies
   * the path that all other paths in the source map are relative to. If this
   * field is not present, all paths in the source map are interpreted as being
   * relative to the directory containing the source map instead.
   *
   * @see https://esbuild.github.io/api/#source-root
   */
  sourceRoot?: URL | string | null | undefined

  /**
   * Include the `sourcesContent` field in sourcemaps.
   *
   * @see https://esbuild.github.io/api/#sources-content
   */
  sourcesContent?: boolean | null | undefined
}

/**
 * Create a plugin to transform files with esbuild.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Plugin}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Options} [opts]
 *  esbuild options
 * @return {Plugin}
 *  esbuild transform plugin
 */
function plugin(this: void, opts: Options): Plugin {
  /**
   * Copied modules.
   *
   * @const {Set<ModuleInfo>} copied
   */
  const copied: Set<ModuleInfo> = new Set<ModuleInfo>()

  /**
   * Module id filter.
   *
   * @const {(id: unknown) => boolean} filter
   */
  const filter: (id: unknown) => boolean = createFilter(
    opts.include ?? /\.(?:(?:json)|(?:[cm]?[jt]s)|(?:[jt]sx?))$/,
    opts.exclude ?? undefined
  )

  /**
   * Raw tsconfig.
   *
   * @var {TsconfigRaw} tsconfigRaw
   */
  let tsconfigRaw: TsconfigRaw

  return {
    buildStart,
    generateBundle,
    name: PLUGIN_NAME,
    onLog,
    renderChunk,
    transform
  }

  /**
   * Get the raw tsconfig from `options`.
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {undefined}
   */
  function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): undefined {
    ok(typeof options.tsconfig === 'object', 'expected `options.tsconfig`')
    tsconfigRaw = options.tsconfig as TsconfigRaw
    return void this
  }

  /**
   * Add the content of copied modules to their chunks. Any sourcemaps generated
   * for copied modules will be removed.
   *
   * @see https://rollupjs.org/plugin-development/#generatebundle
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} options
   *  Normalized output options
   * @param {OutputBundle} bundle
   *  Output bundle
   * @return {undefined}
   */
  function generateBundle(
    this: PluginContext,
    options: NormalizedOutputOptions,
    bundle: OutputBundle
  ): undefined {
    for (const output of Object.values(bundle)) {
      for (const info of copied) {
        if (output.type === 'chunk' && output.facadeModuleId === info.id) {
          ok(typeof info.code === 'string', 'expected `info.code`')

          output.code = info.code
          output.map = null

          if (output.sourcemapFileName !== null) {
            delete bundle[output.sourcemapFileName]
            output.sourcemapFileName = null
          }
        }
      }
    }

    return void bundle
  }

  /**
   * Skip `EMPTY_BUNDLE` logs.
   *
   * @see https://rollupjs.org/plugin-development/#onlog
   *
   * @this {MinimalPluginContext}
   *
   * @param {RollupLogLevel} level
   *  Rollup log level
   * @param {RollupLog} log
   *  Rollup log object
   * @return {false | undefined}
   *  `false` if log will be silenced, `undefined` otherwise
   */
  function onLog(
    this: MinimalPluginContext,
    level: RollupLogLevel,
    log: RollupLog
  ): false | undefined {
    return ifelse(log.code === 'EMPTY_BUNDLE', false, void log)
  }

  /**
   * Fix the name and file name of copied modules.
   *
   * @see https://rollupjs.org/plugin-development/#renderchunk
   *
   * @this {PluginContext}
   *
   * @param {string} code
   *  Chunk content
   * @param {RenderedChunk} chunk
   *  The chunk to process
   * @return {undefined}
   */
  function renderChunk(
    this: PluginContext,
    code: string,
    chunk: RenderedChunk
  ): undefined {
    for (const info of copied) {
      if (chunk.facadeModuleId === info.id) {
        chunk.fileName = pathe.changeExt(chunk.fileName, pathe.extname(info.id))
        chunk.name = pathe.changeExt(chunk.fileName, null)
        break
      }
    }

    return void chunk
  }

  /**
   * Transform `code` using the [esbuild transform api][esbuild-transform].
   *
   * [esbuild-transform]: https://esbuild.github.io/api/#transform
   *
   * @see https://rollupjs.org/plugin-development/#transform
   *
   * @todo `globalName`
   * @todo `ignoreAnnotations`
   * @todo `pure`
   * @todo `treeShaking`
   *
   * @this {TransformPluginContext}
   *
   * @param {string} code
   *  The code to transform
   * @param {string} id
   *  The id of the module containing `code`
   * @return {Promise<TransformResult>}
   *  Transform result
   */
  async function transform(
    this: TransformPluginContext,
    code: string,
    id: string
  ): Promise<TransformResult> {
    /**
     * Transform result.
     *
     * @var {WithUndefined<Partial<SourceDescription>> | undefined} result
     */
    let result: WithUndefined<Partial<SourceDescription>> | undefined

    if (filter(id)) {
      /**
       * AST of transformed code.
       *
       * @var {ProgramNode | undefined} ast
       */
      let ast: ProgramNode | undefined

      /**
       * Content loader.
       *
       * @var {Loader | undefined} loader
       */
      let loader: Loader | undefined = pathe.extToValue(pathe.basename(id), {
        '.cjs': 'js',
        '.cts': 'ts',
        '.d.cts': 'copy',
        '.d.mts': 'copy',
        '.d.ts': 'copy',
        '.js': 'js',
        '.json': 'copy',
        '.jsx': 'jsx',
        '.mjs': 'js',
        '.mts': 'ts',
        '.ts': 'ts',
        '.tsx': 'tsx',
        ...opts.loader
      })

      // note: rather than just be a no-op, or work as `empty`, esbuild throws
      // an annoying error when the `copy` loader is used with `transform`.
      if (loader === 'copy') {
        /**
         * Module info.
         *
         * @const {ModuleInfo | null} info
         */
        const info: ModuleInfo | null = this.getModuleInfo(id)

        if (info) {
          copied.add(info)
          info.code = code
        }

        ast = {
          body: [],
          end: 0,
          sourceType: 'module',
          start: 0,
          type: 'Program'
        }

        loader = 'empty'
      }

      const { transform } = esbuild ??= await import('esbuild')

      /**
       * Log level overrides.
       *
       * @const {Record<string, LogLevel>} logOverride
       */
      const logOverride: Record<string, LogLevel> = { ...opts.logOverride }

      ok(typeof opts.sourcesContent === 'boolean', 'expected `sourcesContent`')

      const {
        code: transformed,
        map,
        legalComments,
        mangleCache,
        warnings
      } = await transform(code, {
        charset: opts.charset ?? undefined,
        color: false,
        define: opts.define ?? undefined,
        drop: opts.drop ?? undefined,
        dropLabels: opts.dropLabels ?? undefined,
        format: opts.format,
        jsx: opts.jsx ?? undefined,
        jsxDev: opts.jsxDev ?? undefined,
        jsxFactory: opts.jsxFactory ?? undefined,
        jsxFragment: opts.jsxFragment ?? undefined,
        jsxImportSource: opts.jsxImportSource ?? undefined,
        jsxSideEffects: opts.jsxSideEffects ?? undefined,
        keepNames: opts.keepNames ?? undefined,
        legalComments: 'external',
        lineLimit: opts.lineLimit ?? undefined,
        loader,
        logLevel: 'silent',
        logLimit: 0,
        logOverride: Object
          .entries(logOverride)
          .reduce<Record<string, LogLevel>>((acc, [key, level]) => {
            acc[key] = ifelse(level === 'silent', level, 'warning')
            return acc
          }, {}),
        mangleCache: opts.mangleCache ?? undefined,
        mangleProps: opts.mangleProps ?? undefined,
        mangleQuoted: opts.mangleQuoted ?? undefined,
        minify: opts.minify ?? undefined,
        minifyIdentifiers: opts.minifyIdentifiers ?? undefined,
        minifySyntax: opts.minifySyntax ?? undefined,
        minifyWhitespace: opts.minifyWhitespace ?? undefined,
        platform: opts.platform ?? undefined,
        reserveProps: opts.reserveProps ?? undefined,
        sourceRoot: ifelse(
          isNIL(opts.sourceRoot),
          undefined,
          String(opts.sourceRoot)
        ),
        sourcefile: id,
        sourcemap: true,
        sourcesContent: opts.sourcesContent,
        supported: {
          'dynamic-import': true,
          'import-meta': true,
          ...opts.supported
        },
        target: typeof opts.target === 'object'
          ? [...new Set(opts.target)]
          : opts.target,
        tsconfigRaw: opts.tsconfigRaw ?? tsconfigRaw
      } as TransformOptions)

      for (const message of warnings) {
        /**
         * Message formatted as rollup log.
         *
         * @const {RollupLog} log
         */
        const log: RollupLog = formatEsbuildMessage(
          message,
          code,
          id,
          PLUGIN_NAME,
          logOverride[message.id]
        )

        // log message if not silenced
        log.level && this[log.level](log, log.loc)
      }

      result = {
        ast,
        code: typeof ast === 'object' ? code : transformed,
        map,
        meta: { legalComments, mangleCache, warnings }
      }
    }

    return result as TransformResult
  }
}
