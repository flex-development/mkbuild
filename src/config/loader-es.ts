/**
 * @file Config - esLoader
 * @module mkbuild/config/esLoader
 */

import type { Config } from '#src/interfaces'
import tsconfigPaths from '#src/plugins/tsconfig-paths/plugin'
import { build } from 'esbuild'
import { resolveImports, toDataURL } from 'mlly'
import { RESOLVE_EXTENSIONS } from './constants'

/**
 * Loads an [ESM][1] (`*.js`, `*.mjs`) or TypeScript (`*.cts`, `*.mts`, `*.ts`)
 * build config file.
 *
 * [1]: https://nodejs.org/api/esm.html
 *
 * @see https://github.com/davidtheclark/cosmiconfig#loaders
 *
 * @async
 *
 * @param {string} path - Config file path
 * @param {string} content - Config file content
 * @return {Promise<Config>} Build configuration options
 */
const esLoader = async (path: string, content: string): Promise<Config> => {
  const {
    outputFiles: [output]
  } = await build({
    entryPoints: [path],
    format: 'esm',
    loader: {
      '.cts': 'ts',
      '.js': 'js',
      '.mjs': 'js',
      '.mts': 'ts',
      '.ts': 'ts'
    },
    metafile: true,
    outdir: '<stdout>',
    plugins: [tsconfigPaths()],
    write: false
  })

  // resolve imports in content to evaluate module
  content = await resolveImports(output!.text, {
    extensions: RESOLVE_EXTENSIONS,
    url: path
  })

  return ((await import(toDataURL(content))) as { default: Config }).default
}

export default esLoader
