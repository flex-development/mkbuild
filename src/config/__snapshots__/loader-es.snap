// Vitest Snapshot v1

exports[`unit:config/esLoader > should return config from .cts file 1`] = `
{
  "absWorkingDir": "__fixtures__/configs/cts",
  "entries": [
    {
      "ext": ".cjs",
      "format": "cjs",
    },
  ],
}
`;

exports[`unit:config/esLoader > should return config from .js file 1`] = `
{
  "absWorkingDir": "__fixtures__/configs/js",
  "entries": [
    {
      "format": "esm",
    },
  ],
}
`;

exports[`unit:config/esLoader > should return config from .mjs file 1`] = `
{
  "absWorkingDir": "__fixtures__/configs/mjs",
  "entries": [
    {
      "format": "esm",
    },
  ],
}
`;

exports[`unit:config/esLoader > should return config from .mts file 1`] = `
{
  "absWorkingDir": "__fixtures__/configs/mts",
  "entries": [
    {
      "format": "esm",
    },
  ],
}
`;

exports[`unit:config/esLoader > should return config from .ts file 1`] = `
{
  "absWorkingDir": "__fixtures__/configs/ts",
  "entries": [
    {
      "format": "esm",
    },
    {
      "ext": ".cjs",
      "format": "cjs",
    },
  ],
}
`;
